### PR TITLE
I3bar json

### PR DESCRIPTION
**Pull request by SirCmpwn**
Test plan:

```
bar {
    status_command i3status
}
```

Bonus: test click events by running the bar manually with `-d` and it'll log each block that's clicked.

### CODE REVIEW

File: swaybar/i3bar.c [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+		json_object_object_get_ex(json, "border_left", &border_left);
+		json_object_object_get_ex(json, "border_right", &border_right);
+
+		struct i3bar_block *block = calloc(1, sizeof(struct i3bar_block));
+		block->full_text = full_text ?
+			strdup(json_object_get_string(full_text)) : NULL;
+		block->short_text = short_text ?
+			strdup(json_object_get_string(short_text)) : NULL;
+		if (color) {
+			block->color = malloc(sizeof(uint32_t));
```
Comment (ascent12) [inline swaybar/i3bar.c]: Allocating a single uint32_t?

File: swaybar/i3bar.c [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+	if (block->instance) {
+		json_object_object_add(event_json, "instance",
+				json_object_new_string(block->instance));
+	}
+
+	json_object_object_add(event_json, "button", json_object_new_int(button));
+	json_object_object_add(event_json, "x", json_object_new_int(x));
+	json_object_object_add(event_json, "y", json_object_new_int(y));
+
+	char event_buf[1024];
```
Comment (ascent12) [inline swaybar/i3bar.c]: You could rid of the need for this by using dprintf instead.
```c
dprintf(write_fd, "%s\n", json_object_to_json_string(event_json));
```

File: swaybar/i3bar.c (hunk 1 above)
Comment (SirCmpwn) [reply swaybar/i3bar.c]: Yeah cause it can be NULL

Comment (SirCmpwn) [reply swaybar/i3bar.c]: 0 != NULL in this case

Comment (ascent12) [reply swaybar/i3bar.c]: I think an accompanying "is_valid" bool would be more appropriate.

Comment (SirCmpwn) [reply swaybar/i3bar.c]: nack

**Comment by emersion** [comment]
Crashes with my config:

<details>
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00005608462d486b in render_status_block (cairo=0x56084850aaa0, config=0x5608484873e0, 
    output=0x560848490580, block=0x560848516200, x=0x7fff0b5aecd0, height=24, focused=true, edge=false)
    at ../swaybar/render.c:104
104		if (!block->full_text || !*block->full_text) {
(gdb) bt
#0  0x00005608462d486b in render_status_block (cairo=0x56084850aaa0, config=0x5608484873e0, output=0x560848490580, block=0x560848516200, x=0x7fff0b5aecd0, height=24, focused=true, edge=false)
    at ../swaybar/render.c:104
#1  0x00005608462d47cb in render_status_line_i3bar (cairo=0x56084850aaa0, config=0x5608484873e0, output=0x560848490580, status=0x560848489ca0, focused=true, x=0x7fff0b5aecd0, width=1024, height=24)
    at ../swaybar/render.c:234
#2  0x00005608462d4205 in render_status_line (cairo=0x56084850aaa0, config=0x5608484873e0, output=0x560848490580, status=0x560848489ca0, focused=true, x=0x7fff0b5aecd0, width=1024, height=24)
    at ../swaybar/render.c:254
#3  0x00005608462d37bf in render_to_cairo (cairo=0x56084850aaa0, bar=0x5608464dc780 <swaybar>, output=0x560848490580) at ../swaybar/render.c:412
#4  0x00005608462d343f in render_frame (bar=0x5608464dc780 <swaybar>, output=0x560848490580)
    at ../swaybar/render.c:437
#5  0x00005608462cf0c3 in layer_surface_configure (data=0x560848490580, surface=0x5608484a8460, serial=91, width=1024, height=24) at ../swaybar/bar.c:44
#6  0x00007f7b3e5811c8 in ffi_call_unix64 () at /usr/lib/libffi.so.6
#7  0x00007f7b3e580c2a in ffi_call () at /usr/lib/libffi.so.6
#8  0x00007f7b43582ddd in wl_closure_invoke (closure=<optimized out>, flags=<optimized out>, target=<optimized out>, opcode=0, data=<optimized out>) at src/connection.c:995
#9  0x00007f7b43581127 in dispatch_event (display=<optimized out>, queue=<optimized out>)
    at src/wayland-client.c:1434
#10 0x00007f7b43580bb9 in dispatch_queue (display=<optimized out>, queue=<optimized out>)
    at src/wayland-client.c:1580
#11 0x00007f7b43580bb9 in wl_display_dispatch_queue_pending (display=0x560848489980, queue=0x560848489a48) at src/wayland-client.c:1822
#12 0x00007f7b43580582 in wl_display_dispatch_queue (display=0x560848489980, queue=0x560848489a48)
    at src/wayland-client.c:1795
#13 0x00007f7b435802ab in wl_display_roundtrip_queue (display=0x560848489980, queue=0x560848489a48)
    at src/wayland-client.c:1245
#14 0x00005608462d34c1 in render_frame (bar=0x5608464dc780 <swaybar>, output=0x560848490580)
    at ../swaybar/render.c:449
#15 0x00005608462cfcfd in render_all_frames (bar=0x5608464dc780 <swaybar>) at ../swaybar/bar.c:243
#16 0x00005608462cfea1 in status_in (fd=5, mask=1, _bar=0x5608464dc780 <swaybar>)
    at ../swaybar/bar.c:319
#17 0x00005608462d0bae in event_loop_poll () at ../swaybar/event_loop.c:122
#18 0x00005608462cfdb1 in bar_run (bar=0x5608464dc780 <swaybar>) at ../swaybar/bar.c:330
#19 0x00005608462d3315 in main (argc=4, argv=0x7fff0b5af438) at ../swaybar/main.c:104
```
</details>

**Comment by SirCmpwn** [comment]
Your config?

**Comment by emersion** [comment]
https://sr.ht/PHIw.config

**Comment by emersion** [comment]
The blocks are printed in the reverse order.

There's also this weird bug: when I put a space right after a FontAwesome icon, it prints an invalid 0x0020 codepoint. Any other character is fine.

**Comment by emersion** [comment]
That 0x0020 thing is actually this bug https://bugs.archlinux.org/task/56899

**Comment by SirCmpwn** [comment]
woot

**Comment by emersion** [comment]
Build failed

**Comment by SirCmpwn** [comment]
Build unfailed

**Pull request merged**